api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java

```diff
@@ -0,0 +1,170 @@
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import org.openmrs.CodedOrFreeText;
+import org.openmrs.Diagnosis;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.api.APIException;
+import org.openmrs.api.DiagnosisService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.parameter.EncounterSearchCriteria;
+import org.openmrs.parameter.EncounterSearchCriteriaBuilder;
+import org.openmrs.util.PrivilegeConstants;
+
+/**
+ * API functionality for Diagnoses.
+ *
+ * @since 2.2
+ */
+public class DiagnosisServiceImpl extends BaseOpenmrsService implements DiagnosisService {
+
+	public static final String DIAGNOSIS_ORDER_PRIMARY = "Primary";
+
+	private DiagnosisDAO dao;
+
+	public void setDiagnosisDAO(DiagnosisDAO dao) {
+		this.dao = dao;
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#save(Diagnosis)
+	 */
+	public Diagnosis save(Diagnosis diagnosis) {
+		if (diagnosis.getId() == null) {
+			Context.requirePrivilege(PrivilegeConstants.ADD_DIAGNOSES);
+		} else {
+			Context.requirePrivilege(PrivilegeConstants.EDIT_DIAGNOSES);
+		}
+		return dao.saveDiagnosis(diagnosis);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#voidDiagnosis(Diagnosis, String)
+	 */
+	public Diagnosis voidDiagnosis(Diagnosis diagnosis, String voidReason) {
+		diagnosis.setVoided(true);
+		diagnosis.setVoidedBy(Context.getAuthenticatedUser());
+		diagnosis.setVoidReason(voidReason);
+		diagnosis.setDateVoided(new Date());
+		return save(diagnosis);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getDiagnosisByUuid(String)
+	 */
+	public Diagnosis getDiagnosisByUuid(String uuid) {
+		return dao.getDiagnosisByUuid(uuid);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getDiagnoses(Patient, Date)
+	 */
+	public List<Diagnosis> getDiagnoses(Patient patient, Date fromDate) {
+
+		EncounterSearchCriteria encounterSearchCriteria = new EncounterSearchCriteriaBuilder().setPatient(patient)
+			.setFromDate(fromDate).createEncounterSearchCriteria();
+		List<Encounter> encounters = Context.getEncounterService().getEncounters(encounterSearchCriteria);
+
+		List<Diagnosis> diagnoses = new ArrayList<Diagnosis>();
+
+		for (Encounter encounter : encounters) {
+			diagnoses.addAll(encounter.getDiagnoses());
+		}
+		return diagnoses;
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getPrimaryDiagnoses(Encounter)
+	 */
+	public List<Diagnosis> getPrimaryDiagnoses(Encounter encounter) {
+		List<Diagnosis> diagnoses = null;
```

Comment:
You probably don't want to have this return null, it should return an empty list worst case scenario.